pkg/restore/restore.go

```diff
@@ -308,6 +375,97 @@ func resolveActions(actions []ItemAction, helper discovery.Helper) ([]resolvedAc
 	return resolved, nil
 }
 
+// namespaceWaiter encapsulates channels for namespaces waiting to terminate, along with a mutex for the map.
+type namespaceWaiter struct {
+	chansLock sync.Mutex
+	chans     map[string]chan *v1.Namespace
+}
+
+func (nsw *namespaceWaiter) getDeleteChan(ns string) (chan *v1.Namespace, bool) {
+	nsw.chansLock.Lock()
+	defer nsw.chansLock.Unlock()
+
+	ch, ok := nsw.chans[ns]
+	if !ok {
+		return nil, ok
+	}
+	return ch, ok
+}
+
+func (nsw *namespaceWaiter) addDeleteChan(ns string) {
+	nsw.chansLock.Lock()
+	defer nsw.chansLock.Unlock()
+
+	nsw.chans[ns] = make(chan *v1.Namespace)
+}
+
+func (nsw *namespaceWaiter) removeDeleteChan(ns string) {
+	nsw.chansLock.Lock()
+	defer nsw.chansLock.Unlock()
+
+	delete(nsw.chans, ns)
+}
+
+func (nsw *namespaceWaiter) waitForDelete(ns string, timeout time.Duration) (*v1.Namespace, error) {
+	deleteChan, ok := nsw.getDeleteChan(ns)
+	if !ok {
+		return nil, errors.Errorf("Namespace %s not found in list of channels waiting to delete", ns)
+	}
+
+	select {
+	case <-time.After(timeout):
+		return nil, errors.Errorf("Namespace %s did not terminate before the alloted timeout.", ns)
+	case res := <-deleteChan:
+		return res, nil
+	}
+}
+
+// pvWaiter encapsulates waiting for a PV with a reclaim policy of 'Delete' to exit the 'Released' state.
+// This is to avoid the condition where a PV associated with a PV is released but not yet deleted not getting restored.
+type pvWaiter struct {
```

Comment:
I think we could probably eliminate the need for a second PV-specific waiter type.  Within the waiter struct, we really only need the ObjectMeta; it doesn't really matter what specific type it is.  I'd probably use the `metav1.Object` interface - so if you change the channel type to use that, then I think you can consolidate to a single `resourceWaiter` struct.

Reply:
Ok, that makes sense.